File: backend/app/scraping/pipeline.py
import hashlib

import structlog
from tenacity import retry, stop_after_attempt, wait_exponential

from app.models.property import PropertyIn
from app.repositories.properties import PropertyRepository
from app.scraping.sources.base import PropertySourceAdapter, ScrapeContext

logger = structlog.get_logger(__name__)


class ScrapingPipeline:
    def __init__(self, repository: PropertyRepository, adapters: list[PropertySourceAdapter]):
        self.repository = repository
        self.adapters = adapters

    async def run(self, context: ScrapeContext) -> dict[str, int]:
        inserted = 0
        for adapter in self.adapters:
            listings = await self._fetch_with_retry(adapter, context)
            for listing in listings:
                normalized = self._normalize(listing)
                await self.repository.upsert_by_dedupe_key(normalized.model_dump())
                inserted += 1
            logger.info("scrape_source_completed", source=adapter.source_name, count=len(listings))
        return {"processed": inserted}

    @retry(wait=wait_exponential(multiplier=1, min=2, max=30), stop=stop_after_attempt(3))
    async def _fetch_with_retry(self, adapter: PropertySourceAdapter, context: ScrapeContext) -> list[PropertyIn]:
        return await adapter.fetch(context)

    def _normalize(self, listing: PropertyIn) -> PropertyIn:
        if listing.dedupe_key:
            return listing
        key = hashlib.sha256(f"{listing.title}:{listing.rent}:{listing.location.coordinates}".encode()).hexdigest()
        return listing.model_copy(update={"dedupe_key": key})

File: backend/src/scraping/pipeline.ts
import crypto from "crypto";
import { logger } from "../core/logger.js";
import { PropertyBase } from "../models/property.js";
import { PropertyRepository } from "../repositories/properties.js";
import { PropertySourceAdapter, ScrapeContext } from "./sources/base.js";

export class ScrapingPipeline {
  private repository: PropertyRepository;
  private adapters: PropertySourceAdapter[];

  constructor(repository: PropertyRepository, adapters: PropertySourceAdapter[]) {
    this.repository = repository;
    this.adapters = adapters;
  }

  async run(context: ScrapeContext): Promise<{ processed: number }> {
    let inserted = 0;
    for (const adapter of this.adapters) {
      try {
        const listings = await this.fetchWithRetry(adapter, context);
        for (const listing of listings) {
          const normalized = this.normalize(listing);
          await this.repository.upsertByDedupeKey(normalized);
          inserted += 1;
        }
        logger.info(
          { source: adapter.sourceName, count: listings.length },
          "scrape_source_completed"
        );
      } catch (error) {
        logger.error(
          { source: adapter.sourceName, error: (error as Error).message },
          "scrape_source_failed"
        );
      }
    }
    return { processed: inserted };
  }

  private async fetchWithRetry(
    adapter: PropertySourceAdapter,
    context: ScrapeContext,
    maxRetries: number = 3
  ): Promise<PropertyBase[]> {
    let attempt = 0;
    let delay = 1000;

    while (attempt < maxRetries) {
      try {
        return await adapter.fetch(context);
      } catch (err) {
        attempt++;
        if (attempt >= maxRetries) {
          throw err;
        }
        await new Promise((resolve) => setTimeout(resolve, delay));
        delay *= 2;
      }
    }
    return [];
  }

  private normalize(listing: PropertyBase): Record<string, any> {
    if (listing.dedupe_key) {
      return { ...listing };
    }
    const coords = listing.location?.coordinates || [0, 0];
    const key = crypto
      .createHash("sha256")
      .update(`${listing.title}:${listing.rent}:${coords[0]},${coords[1]}`)
      .digest("hex");

    return {
      ...listing,
      dedupe_key: key,
    };
  }
}
